File: app/src/main/java/com/android/retrofitsampleapp2/iu/users/GitUsersAdapter.java
package com.android.retrofitsampleapp2.iu.users;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.android.retrofitsampleapp2.R;
import com.android.retrofitsampleapp2.domain.GitUserEntity;

import java.util.List;

public class GitUsersAdapter extends RecyclerView.Adapter<GitUsersViewHolder> {

    private List<GitUserEntity> data;
    private OnItemClickListener listener;//завели слушатель

    public void setData(List<GitUserEntity> users) {
        data = users;
        notifyDataSetChanged();
    }

    //сохнанили слушатель. после интерфейса идем сюда и передаем его дальше во onCreateViewHolder
    public void setOnItemClickListener(OnItemClickListener listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public GitUsersViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new GitUsersViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_git_user, parent, false), listener);// listener - передаем далее во viewHolder
    }

    @Override
    public void onBindViewHolder(@NonNull GitUsersViewHolder holder, int position) {
        holder.bind(getItem(position));
    }

    private GitUserEntity getItem(int pos) {
        return data.get(pos);
    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    //интерфейс со слушателем
    public interface OnItemClickListener {
        void onItemClick(GitUserEntity user);
    }
}

File: app/src/main/java/com/android/retrofitsampleapp2/iu/users/GitUsersAdapter.kt
package com.android.retrofitsampleapp2.iu.users

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.android.retrofitsampleapp2.R
import com.android.retrofitsampleapp2.domain.GitUserEntity

class GitUsersAdapter(
    private var data: List<GitUserEntity>,
    private var listener: (GitUserEntity) -> Unit
) : RecyclerView.Adapter<GitUsersViewHolder>() {

    @SuppressLint("NotifyDataSetChanged")
    fun setData(users: List<GitUserEntity>) {
        data = users
        notifyDataSetChanged()
    }

    //сохнанили слушатель. после интерфейса идем сюда и передаем его дальше во onCreateViewHolder
    fun setOnItemClickListener(listener: (GitUserEntity) -> Unit) {
        this.listener = listener
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): GitUsersViewHolder {
        return GitUsersViewHolder(
            LayoutInflater.from(parent.context)
                .inflate(R.layout.item_git_user, parent, false),
            (listener as OnItemClickListener)
        ) // listener - передаем далее во viewHolder
    }

    override fun onBindViewHolder(holder: GitUsersViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    private fun getItem(pos: Int): GitUserEntity = data[pos]

    override fun getItemCount(): Int = data.size

    //интерфейс со слушателем
    interface OnItemClickListener : (OnItemClickListener) -> Unit {
        fun onItemClick(user: GitUserEntity)
    }
}